contracts/uniswap/UniswapReserve.sol

```diff
@@ -0,0 +1,300 @@
+        uint256 min_tokens,
+        uint256 deadline
+    )
+        external
+        payable
+        returns (uint256  tokens_bought);
+
+    function tokenToEthSwapInput(
+        uint256 tokens_sold,
+        uint256 min_eth,
+        uint256 deadline
+    )
+        external
+        returns (uint256  eth_bought);
+}
+
+
+interface UniswapFactory {
+    function getExchange(address token) external view returns (address exchange);
+}
+
+
+contract UniswapReserve is KyberReserveInterface, Withdrawable, Utils2 {
+    // Parts per 10000
+    uint public constant DEFAULT_FEE_BPS = 25;
+
+    UniswapFactory public uniswapFactory;
+    address public kyberNetwork;
+
+    uint public feeBps = DEFAULT_FEE_BPS;
+
+    // Uniswap exchange contract for every listed token
+    // token -> exchange
+    mapping (address => address) public tokenExchange;
+
+    bool public tradeEnabled = true;
+
+    /**
+        Constructor
+    */
+    function UniswapReserve(
+        UniswapFactory _uniswapFactory,
+        address _admin,
+        address _kyberNetwork
+    )
+        public
+    {
+        require(address(_uniswapFactory) != 0);
+        require(_admin != 0);
+        require(_kyberNetwork != 0);
+
+        uniswapFactory = _uniswapFactory;
+        admin = _admin;
+        kyberNetwork = _kyberNetwork;
+    }
+
+    function() public payable {
+        // anyone can deposit ether
+    }
+
+    /**
+        Returns dest quantity / source quantity.
+    */
+    function getConversionRate(
+        ERC20 src,
+        ERC20 dest,
+        uint srcQty,
+        uint blockNumber
+    )
+        public
+        view
+        returns(uint)
+    {
+        // This makes the UNUSED warning go away.
+        blockNumber;
+
+        require(isValidTokens(src, dest));
+
+        if (!tradeEnabled) return 0;
+
+        ERC20 token;
+        if (src == ETH_TOKEN_ADDRESS) {
+            token = dest;
+        } else if (dest == ETH_TOKEN_ADDRESS) {
+            token = src;
+        } else {
+            // Should never arrive here - isValidTokens requires one side to be ETH
+            revert();
+        }
+
+        UniswapExchange exchange = UniswapExchange(tokenExchange[token]);
+
+        uint convertedQuantity;
+        if (src == ETH_TOKEN_ADDRESS) {
+            uint quantity = srcQty * (10000 - feeBps) / 10000;
+            convertedQuantity = exchange.getEthToTokenInputPrice(quantity);
+        } else {
+            convertedQuantity = exchange.getTokenToEthInputPrice(srcQty);
+            convertedQuantity = convertedQuantity * (10000 - feeBps) / 10000;
+        }
+
+        return calcRateFromQty(
+            srcQty, /* srcAmount */
+            convertedQuantity, /* destAmount */
+            getDecimals(src), /* srcDecimals */
+            getDecimals(dest) /* dstDecimals */
+        );
+    }
+
+    event TradeExecute(
+        address indexed sender,
+        address src,
+        uint srcAmount,
+        address destToken,
```

Comment:
for src its src only.
for dest its destToken

better be the same
of is it different in Kyber classic reserve?